core/capabilities.go

```diff
@@ -723,6 +715,19 @@ func (bcast *Capabilities) LegacyOnly() bool {
 	return bcast.bitstring.CompatibleWith(legacyCapabilityString)
 }
 
+func (bcast *Capabilities) SetPerCapabilityConstraints(constraints PerCapabilityConstraints) {
+	if bcast != nil {
+		bcast.constraints.perCapability = constraints
```

Comment:
I think your nil reference might've been because the `perCapability` map is not being initialized? You can't write to a map in go without first constructing one (differently from slices).

So perhaps the simplest here would be to add a logic to check if the `perCapability` is `nil` and initializing it so, wdyt?

```suggestion
		if bcast.constraints.perCapability == nil {
			bcast.constraints.perCapability = PerCapabilityConstraints{}
		}
		bcast.constraints.perCapability = constraints
```

Reply:
@victorges, thanks for your comment. I investigated the issue, and while this problem wasn't the cause of the `nil` error I experienced, I agree that this `nil` error can occur. To address it, I now initialize the field in the `NewCapabilities` constructor.

The cause of the `nil` error I encountered was due to Gateways with old software, which didn't have the RPC fields, occasionally calling my orchestrator during debugging. This explained the infrequency of the crash. I've now handled it by checking if the new fields are present (see [commit f442382](https://github.com/livepeer/go-livepeer/pull/3115/commits/f44238233b09424bb1fde2e1133d5bd08a23b925)).